### DISCUSSION TITLE
General question about underline color ANSI escape codes

### DISCUSSION THREAD

**Discussion by unphased** [kovidgoyal/kitty#6473] [Q&A]
I have seen this page https://sw.kovidgoyal.net/kitty/underlines/ which people seem to be referring to as the "spec". 

I have two main questions. 

1. Question 1 is about the use of colons in the `\e[58m` sequences. Clearly from this page it is explained that colons are used to select the underline style for `\e[4m` underline specification. This must be the case, as `\e[4;2m` would indeed tell the terminal to set underline and dim, rather than double underline. **However,** in places like [here](https://github.com/microsoft/vscode/issues/158827) we see it stated that colons are also used for the `\e[58m` sequences. My problem with this is that I do not see any examples showing use of colons for specifying the 256 color and true color colors for `\e[38m` `\e[48m` sequences. it's always semicolons. I also tested and see that for `\e[58m` both colons and semicolons work. So my question is **which one is the standard for `\e[58m`, colons or semicolons**?
2. ~~I like many others customize my 8 standard terminal colors (black, red, green, yellow, blue, magenta, cyan, white) to a palette of careful choosing, for eye-pleasing colors. The typical default values of e.g. #000, #f00, #0f0, etc are very garish on the eyes. What's sorely missing from the underline color spec is a range of codes to specify the 8-color colors for underline. I tried `\e[50m` thru `\e[57m` to no avail. The problem is that although I can get the sequence (black, red, green etc) out of `\e[58;5;1m` thru `\e[58;5;7m`, these colors are not being applied with the color palette specified for the terminal emulator.~~ OK I realized that this is actually really just a bug in alacritty. I see that Wezterm gets this correct. After all, with alacritty it does spit out the correct palette colors for `\e[48;5;1~7m` so it is just an oversight in the underline color implementation. AFAIK this is the only way to get color palette bgcolors anyhow.

Sorry for posting a question not actually specific to the kitty terminal.

**Comment by kovidgoyal** [comment]
1. colons: semi-colons must never be used for colors, their existence is
an abomination perpetrated by Konsole making it impossible to combine
SGR codes.

2. I have no interest in/knowledge of alacritty bugs

**Reply by unphased** [reply]
Thanks, and yes i only left the text for the 2nd thing there after i solved it on my own (rubber ducking here) since I figure it probably would help the other 2 people on earth who might be digging around in this same rabbit hole as me.

OK, based on what you just said I can confirm that colons work well in all terminals except macOS builtin terminal.app. That one fails to apply 256color bgcolor via `\e[48:5:<ID>m` with colons and succeeds with semicolons. 

I don't use that terminal obviously, but this makes it tricky, because there are going to be a nontrivial number of potential users of any CLI app that may not know any better yet and are still using that terminal, so we're pushed toward using semicolons here for better compatibility. Do you have any thoughts on this?

**Comment by kovidgoyal** [comment]
On Sat, Jul 15, 2023 at 10:37:23PM -0700, Steven Lu wrote:
> I don't use that terminal obviously but this makes it tricky because there are going to be a nontrivial number of potential users of any CLI app that may use that terminal so we're pushed toward using semicolons here for better compatibility. Do you have any thoughts on this? 

If I were you I would just tell my users to avoid Terminal.app, this is
far from the only problem you will find with it. IIRC it doesnt even
implement basic parsing of escape code types such as DCS escape codes.
It was basically developed many years ago and then put out to pasture by
Apple.
